x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -173,69 +175,123 @@ private AuthorizationInfo getAuthorizationInfoFromContext() {
-            if (operatorException != null) {
-                listener.onFailure(denialException(authentication, action, originalRequest, operatorException));
+            if (checkOperatorPrivileges(authentication, action, originalRequest, listener) == false) {
                 return;
             }
-            operatorPrivilegesService.maybeInterceptRequest(threadContext, originalRequest);
 
             if (SystemUser.is(authentication.getUser())) {
                 // this never goes async so no need to wrap the listener
                 authorizeSystemUser(authentication, action, auditId, unwrappedRequest, listener);
             } else {
-                final String finalAuditId = auditId;
-                final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action);
+                final RequestInfo requestInfo = new RequestInfo(authentication, unwrappedRequest, action, parentContext);
+                final AuthorizationEngine engine = getAuthorizationEngine(authentication);
                 final ActionListener<AuthorizationInfo> authzInfoListener = wrapPreservingContext(ActionListener.wrap(
                         authorizationInfo -> {
                             threadContext.putTransient(AUTHORIZATION_INFO_KEY, authorizationInfo);
-                            maybeAuthorizeRunAs(requestInfo, finalAuditId, authorizationInfo, listener);
+                            maybeAuthorizeRunAs(requestInfo, auditId, authorizationInfo, listener);
                         }, listener::onFailure), threadContext);
-                getAuthorizationEngine(authentication).resolveAuthorizationInfo(requestInfo, authzInfoListener);
+                engine.resolveAuthorizationInfo(requestInfo, authzInfoListener);
+            }
+        }
+    }
+
+    @Nullable
+    private static AuthorizationContext extractAuthorizationContext(ThreadContext threadContext, String childAction) {
+        final String parentAction = threadContext.getTransient(ORIGINATING_ACTION_KEY);
+        if (Strings.isNullOrEmpty(parentAction)) {
+            // No parent action
+            return null;
+        }
+        AuthorizationInfo authorizationInfo = threadContext.getTransient(AUTHORIZATION_INFO_KEY);
+        if (authorizationInfo == null) {
+            throw internalError(
+                "Attempt to authorize child action [" + childAction + "] of [" + parentAction + "] without existing authorization info"
+            );
+        }
+        if (isIndexAction(parentAction)) {
+            final IndicesAccessControl parentAccessControl = threadContext.getTransient(AuthorizationServiceField.INDICES_PERMISSIONS_KEY);
+            return new AuthorizationContext(parentAction, authorizationInfo, new IndexAuthorizationResult(true, parentAccessControl));
+        } else {
+            // We assume the parent is true here, but maybe we actually need to store the AuthorizationResult in a transient
```

Comment:
The `AuthorizationResult` doesn't serve much purpose and `IndexAuthorizationResult` is never passed in place of an `AuthorizationResult`. They are not inter-changeable, they could very well be different classes.
In order to preserve this meaning, I would suggest we get the `AuthorizationServiceField.INDICES_PERMISSIONS_KEY` from the thread context on this if branch too (and change the signature of the `AuthorizationContext` to take an `IndexAuthorizationResult` instead of an `AuthorizationResult`), even though it's not going to be used - but tastes vary (can also be null).

Reply:
> shouldn't it already failed and won't reach here

Yes, which is why I felt comfortable creating a `granted` AuthorizationResult, here.

But to @albertzaharovits's point, this code assumes that there are only 2 `AuthorizationResult` types (Index or non-index) and that it is OK to simply create a standard `AuthorizationResult` for anything that isn't an indices action.

- If that assumption is _true_ (as it is today) then this `AuthorizationResult` is useless. It does absolutely nothing.
- If things change so that the assumption is broken then this code will be wrong. 

So, maybe we should just always use a `IndexAuthorizationResult` or even (nullable) `IndicesAccessControl`

Reply:
I've gone with @albertzaharovits's suggestion and just made this an IndicesAccessControl. It's not a perfect  model, but it's clear and does exactly what we need, which I think is a better tradeoff.